numpy/core/include/numpy/ndarraytypes.h

```diff
@@ -1677,6 +1677,12 @@ PyArray_CLEARFLAGS(PyArrayObject *arr, int flags)
     ((PyArrayObject_fields *)arr)->flags &= ~flags;
 }
 
+static NPY_INLINE NPY_RETURNS_BORROWED_REF PyObject *
+PyArray_HANDLER(PyArrayObject *arr)
```

Comment:
Should this be public API?  Otherwise can we move it or at least hide it behind the "internal" define to be clear about it?

I actually also wonder if we should call it `arr->_mem_handler` to at least ask people to not access it directly (more guard possible, but probably not necessary/useful)?

Reply:
`PyArray_HANDLER` is currently the only API to **get** (not set) an array's `mem_handler`. Exposing it in the public API will make advanced users' lives easier (don't make them access it directly :). It's no more risky than e.g. `PyArray_BASE`.

**Actual use-cases**:

1. Find the "true" array allocator (the one that alloacated the underlying data).
```c
while (arr != NULL && PyArray_Check(arr)) {
	if (PyArray_CHKFLAGS((PyArrayObject *) arr, NPY_ARRAY_OWNDATA)) {
		PyObject *_handler_ = PyArray_HANDLER((PyArrayObject *) arr);
		if (!_handler_) {
			PyErr_SetString(PyExc_RuntimeError, "no memory handler found but OWNDATA flag set");
			return -1;
		}
		PyDataMem_Handler *handler = (PyDataMem_Handler *) PyCapsule_GetPointer(_handler_, "mem_handler");
		if (!handler) {
			return -1;
		}
		printf("### %s ###", handler->name);
		return 0;
	}
	arr = PyArray_BASE((PyArrayObject *) arr);
}
PyErr_SetString(PyExc_ValueError, "argument must be an ndarray");
return -1;
```

2. Allocate a new array with another's allocator.
```c
PyObject *a_arr_handler = PyArray_HANDLER(a_arr);
if (!a_arr_handler) {
	PyErr_SetString(PyExc_RuntimeError, "no memory handler found");
	return -1;
}
PyObject *old_handler = PyDataMem_SetHandler(a_arr_handler);
Py_DECREF(a_arr_handler);
if (!old_handler) {
	return -1;
}

// construct b_arr

if (!PyDataMem_SetHandler(old_handler)) {
	Py_DECREF(old_handler);
	return -1;
}
Py_DECREF(old_handler);

// a_arr and b_arr are handled by the same allocator

return 0;
```

Reply:
> It's no more risky than e.g. PyArray_BASE

OK, yeah, we do want to be able to fetch the allocator.  I guess the question I have is then whether we want to define the `PyObject *` that is returned as actually opaque?  Do we consider the fact that this is currently a capsule to be stable ABI?  I suppose it is stable enough, in the sense that it is not terrible if users get the content, since it will at least fail graciously if things change.

Reply:
> OK, yeah, we do want to be able to fetch the allocator. I guess the question I have is then whether we want to define the `PyObject *` that is returned as actually opaque? Do we consider the fact that this is currently a capsule to be stable ABI? I suppose it is stable enough, in the sense that it is not terrible if users get the content, since it will at least fail graciously if things change.

I guess, since this is the same `PyObject` representation that also `PyDataMem_GetHandler`/`PyDataMem_SetHandler` return/accept it has to be stable ABI. The `PyDataMem_Handler`struct that is encapsulated, on the other hand, could be "versioned". Is that what you mean?

If yes, a nice way to do it could be through the capsule name. For example:

user:
```c
handler_capsule = PyCapsule_New(&my_handler, "v1", destructor);
```

numpy:
```c
if (!PyCapsule_IsValid(mem_handler, "v1") {
   // get pointer and cast it as PyDataMem_Handler_v1
} else if (!PyCapsule_IsValid(handler_capsule, "v2") {
  // get pointer and cast it as PyDataMem_Handler_v2
} else {
  // unknown version
}
```